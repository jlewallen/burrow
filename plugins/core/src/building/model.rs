use crate::library::model::*;

use std::str::FromStr;

#[derive(Clone, Serialize, Deserialize, PartialEq, ToTaggedJson, Reply, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Constructed {
    Area(AreaObservation),
}

impl TryFrom<Constructed> for Effect {
    type Error = TaggedJsonError;

    fn try_from(value: Constructed) -> std::result::Result<Self, Self::Error> {
        Ok(Self::Reply(value.to_tagged_json()?.into()))
    }
}

#[derive(Debug, Serialize, ToTaggedJson)]
#[serde(rename_all = "camelCase")]
struct EditorReply {}

impl Reply for EditorReply {}

#[derive(Default, Clone, Debug)]
pub struct QuickEdit {
    pub name: Option<String>,
    pub desc: Option<String>,
}

impl TryFrom<&EntityPtr> for QuickEdit {
    type Error = DomainError;

    fn try_from(value: &EntityPtr) -> std::result::Result<Self, Self::Error> {
        let name = Some(value.name()?);
        let desc = value.desc()?;
        Ok(Self { name, desc })
    }
}

const SEPARATOR: &str =
    "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~";

impl FromStr for QuickEdit {
    type Err = DomainError;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts: Vec<_> = s.split(SEPARATOR).collect();
        if parts.len() != 2 {
            return Err(DomainError::Anyhow(anyhow::anyhow!("malformed quick edit")));
        }

        let (name, desc) = match parts[..] {
            [name, desc] => (name, desc),
            _ => todo!(),
        };

        let name = Some(name.trim().to_owned());
        let desc = Some(desc.trim().to_owned());

        Ok(Self { name, desc })
    }
}

impl ToString for QuickEdit {
    fn to_string(&self) -> String {
        format!(
            "{}\n\n{}\n\n{}",
            self.name.as_deref().unwrap_or(""),
            SEPARATOR,
            self.desc.as_deref().unwrap_or("")
        )
    }
}
