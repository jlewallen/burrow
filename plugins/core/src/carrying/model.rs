use crate::{library::model::*, tools};

pub use kernel::common::Carrying;

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Containing {
    pub(crate) holding: Vec<EntityRef>,
    pub(crate) capacity: Option<u32>,
    pub(crate) produces: HashMap<String, String>,
}

impl Scope for Containing {
    fn scope_key() -> &'static str {
        "containing"
    }
}

impl Containing {
    pub fn start_carrying(&mut self, item: &EntityPtr) -> Result<bool, DomainError> {
        if let Some(carryable) = item.scope::<Carryable>()? {
            let holding = self
                .holding
                .iter()
                .map(|h| h.to_entity())
                .collect::<Result<Vec<_>, _>>()?;

            for held in holding {
                if is_kind(&held, &carryable.kind)? {
                    let mut combining = held.scope_mut::<Carryable>()?;

                    combining.increase_quantity(&carryable.quantity.into())?;

                    combining.save()?;

                    get_my_session()?.obliterate(item)?;

                    return Ok(true);
                }
            }
        }

        if !self.is_holding(item) {
            self.holding.push(item.entity_ref());
        }

        Ok(true)
    }

    pub fn is_holding(&self, item: &EntityPtr) -> bool {
        self.holding.iter().any(|i| *i.key() == item.key())
    }

    fn remove_item(&mut self, item: &EntityPtr) -> Result<bool, DomainError> {
        self.holding = self
            .holding
            .iter()
            .flat_map(|i| {
                if *i.key() == item.key() {
                    vec![]
                } else {
                    vec![i.clone()]
                }
            })
            .collect::<Vec<EntityRef>>()
            .to_vec();

        Ok(true)
    }

    pub fn stop_carrying(&mut self, found: Found) -> Result<Option<EntityPtr>, DomainError> {
        let item = found.entity()?;

        if !self.is_holding(item) {
            return Ok(None);
        }

        if let Some(carryable) = item.scope::<Carryable>()? {
            match &found {
                Found::One(item) => {
                    if carryable.quantity > 1.0 {
                        let (_original, separated) = tools::separate(item, &1.0.into())?;

                        Ok(Some(separated))
                    } else {
                        self.remove_item(item)?;

                        Ok(Some(item.clone()))
                    }
                }
                Found::Quantified(q, entity) => {
                    if carryable.quantity > q.as_f32() {
                        let (_original, separated) = tools::separate(entity, q)?;

                        Ok(Some(separated))
                    } else {
                        Ok(None)
                    }
                }
            }
        } else {
            self.remove_item(item)?;

            Ok(Some(item.clone()))
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Carryable {
    kind: Kind,
    quantity: f32,
}

fn is_kind(entity: &EntityPtr, kind: &Kind) -> Result<bool, DomainError> {
    if let Some(carryable) = entity.scope::<Carryable>()? {
        Ok(*carryable.kind() == *kind)
    } else {
        Ok(false)
    }
}

impl Default for Carryable {
    fn default() -> Self {
        let session = get_my_session().expect("No session in Entity::new_blank!");
        Self {
            kind: Kind::new(session.new_identity()),
            quantity: 1.0,
        }
    }
}

impl Carryable {
    pub fn quantity(&self) -> f32 {
        self.quantity
    }

    pub fn decrease_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.sanity_check_quantity();

        if *q < 1.0.into() || *q > self.quantity.into() {
            Err(DomainError::Impossible)
        } else {
            self.quantity -= q.as_f32();

            Ok(self)
        }
    }

    pub fn increase_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.sanity_check_quantity();

        self.quantity += q.as_f32();

        Ok(self)
    }

    pub fn set_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.quantity = q.as_f32();

        Ok(self)
    }

    pub fn kind(&self) -> &Kind {
        &self.kind
    }

    pub fn set_kind(&mut self, kind: &Kind) {
        self.kind = kind.clone();
    }

    // Migrate items that were initialized with 0 quantities.
    fn sanity_check_quantity(&mut self) {
        if self.quantity < 1.0 {
            self.quantity = 1.0
        }
    }
}

impl Scope for Carryable {
    fn scope_key() -> &'static str {
        "carryable"
    }
}
