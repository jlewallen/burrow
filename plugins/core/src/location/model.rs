use crate::library::model::*;

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Location {
    pub container: Option<EntityRef>,
}

impl Location {
    pub fn set(item: &EntityPtr, container: EntityRef) -> Result<(), DomainError> {
        let mut location = item.scope_mut::<Location>()?;
        location.container = Some(container);
        location.save()
    }

    pub fn get(item: &EntityPtr) -> Result<Option<EntityRef>, DomainError> {
        match item.scope::<Location>()? {
            Some(location) => Ok(location.container.clone()),
            None => Ok(None),
        }
    }
}

impl Scope for Location {
    fn scope_key() -> &'static str {
        "location"
    }
}

pub fn change_location<A, B, C, D>(
    from: &EntityPtr,
    to: &EntityPtr,
    item: &EntityPtr,
    do_from: C,
    do_into: D,
) -> Result<bool, DomainError>
where
    A: Scope + Serialize,
    B: Scope + Serialize,
    C: FnOnce(&mut A, EntityPtr) -> Result<Option<EntityPtr>, DomainError>,
    D: FnOnce(&mut B, EntityPtr) -> Result<Option<EntityPtr>, DomainError>,
{
    info!("moving {:?} {:?} {:?}", item, from, to);

    let mut from = from.scope_mut::<A>()?;
    let mut into = to.scope_mut::<B>()?;

    match do_from(&mut from, item.clone())? {
        Some(moving) => match do_into(&mut into, moving)? {
            Some(moving) => {
                Location::set(&moving, to.entity_ref())?;
                from.save()?;
                into.save()?;

                Ok(true)
            }
            None => Ok(false),
        },
        None => Ok(false),
    }
}

pub fn container_of(item: &EntityPtr) -> Result<EntityPtr, DomainError> {
    Location::get(item)?
        .ok_or(DomainError::ContainerRequired)?
        .to_entity()
}
