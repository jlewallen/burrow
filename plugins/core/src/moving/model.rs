use crate::library::model::*;

pub trait BeforeMovingHook {
    fn before_moving(&self, surroundings: &Surroundings, to_area: &EntityPtr) -> Result<CanMove>;
}

impl BeforeMovingHook for MovingHooks {
    fn before_moving(&self, surroundings: &Surroundings, to_area: &EntityPtr) -> Result<CanMove> {
        Ok(self
            .before_moving
            .instances
            .borrow()
            .iter()
            .map(|h| h.before_moving(surroundings, to_area))
            .collect::<Result<Vec<CanMove>>>()?
            .iter()
            .fold(CanMove::default(), |c, h| c.fold(h)))
    }
}

pub trait AfterMoveHook {
    fn after_move(&self, surroundings: &Surroundings, from_area: &EntityPtr) -> Result<()>;
}

impl AfterMoveHook for MovingHooks {
    fn after_move(&self, surroundings: &Surroundings, from_area: &EntityPtr) -> Result<()> {
        self.after_move
            .instances
            .borrow()
            .iter()
            .map(|h| h.after_move(surroundings, from_area))
            .collect::<Result<Vec<()>>>()?;

        Ok(())
    }
}

#[derive(Default)]
pub struct MovingHooks {
    pub before_moving: Hooks<Box<dyn BeforeMovingHook>>,
    pub after_move: Hooks<Box<dyn AfterMoveHook>>,
}

impl HooksSet for MovingHooks {
    fn hooks_key() -> &'static str
    where
        Self: Sized,
    {
        "moving"
    }
}

#[derive(Clone, Default)]
pub enum CanMove {
    #[default]
    Allow,
    Prevent,
}

impl HookOutcome for CanMove {
    fn fold(&self, other: &Self) -> Self {
        match (self, other) {
            (_, CanMove::Prevent) => CanMove::Prevent,
            (CanMove::Prevent, _) => CanMove::Prevent,
            (_, _) => CanMove::Allow,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Occupying {
    pub area: EntityRef,
}

impl Scope for Occupying {
    fn scope_key() -> &'static str {
        "occupying"
    }
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SimpleRoute {
    name: String,
    to: EntityRef,
}

impl SimpleRoute {
    pub fn new(name: &str, to: EntityRef) -> Self {
        Self {
            name: name.to_owned(),
            to,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
#[serde(rename_all = "camelCase")]
pub enum Route {
    Simple(SimpleRoute),
}

impl Route {
    fn conflicts_with(&self, other: &Route) -> bool {
        match (self, other) {
            (Route::Simple(a), Route::Simple(b)) => a.name == b.name,
        }
    }

    fn matching_name(&self, name: &str) -> bool {
        match self {
            Route::Simple(simple) => simple.name.to_lowercase().contains(&name.to_lowercase()),
        }
    }

    pub fn destination(&self) -> &EntityRef {
        match self {
            Route::Simple(simple) => &simple.to,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Occupyable {
    pub acls: Acls,
    pub occupied: Vec<EntityRef>,
    pub occupancy: u32,
    pub routes: Option<Vec<Route>>,
}

impl Occupyable {
    pub fn stop_occupying(&mut self, item: &EntityPtr) -> Result<DomainOutcome> {
        let before = self.occupied.len();
        self.occupied.retain(|i| *i.key() != item.key());
        let after = self.occupied.len();
        if before == after {
            return Ok(DomainOutcome::Nope);
        }

        Ok(DomainOutcome::Ok)
    }

    pub fn start_occupying(&mut self, item: &EntityPtr) -> Result<DomainOutcome> {
        self.occupied.push(item.entity_ref());

        Ok(DomainOutcome::Ok)
    }

    pub fn remove_route(&mut self, name: &str) -> Result<bool> {
        if let Some(routes) = &mut self.routes {
            if let Some(found) = routes.iter().position(|r| r.matching_name(name)) {
                routes.remove(found);
                return Ok(true);
            }
        }

        Ok(false)
    }

    pub fn add_route(&mut self, route: Route) -> Result<(), DomainError> {
        let routes = self.routes.get_or_insert_with(|| Vec::new());

        if let Some(conflict) = routes.iter().position(|r| r.conflicts_with(&route)) {
            routes.remove(conflict);
        }

        routes.push(route);

        Ok(())
    }

    pub fn find_route(&self, item: &Item) -> Result<Option<EntityPtr>, DomainError> {
        let Some(routes) = &self.routes else {
            return Ok(None);
        };

        match item {
            Item::Route(name) => {
                for route in routes {
                    if route.matching_name(&name) {
                        return Ok(Some(route.destination().to_entity()?));
                    }
                }
                Ok(None)
            }
            _ => panic!("Occupyable::find_route expecting Item::Route"),
        }
    }
}

impl Scope for Occupyable {
    fn scope_key() -> &'static str {
        "occupyable"
    }
}
