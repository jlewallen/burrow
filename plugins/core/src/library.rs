pub mod plugin {
    pub use anyhow::Result;
    pub use serde::Deserialize;
    pub use tracing::*;

    pub use super::parser::try_parsing;
    pub use kernel::common::DeserializeTagged;
    pub use kernel::prelude::*;
    pub use kernel::try_deserialize_all;
}

pub mod parser {
    use nom::sequence::delimited;
    pub use nom::{
        branch::alt,
        bytes::complete::{tag, take_while, take_while1},
        character::complete::digit1,
        combinator::map,
        combinator::{map_res, opt, recognize},
        sequence::{pair, preceded, separated_pair, terminated, tuple},
        IResult,
    };
    pub use tracing::*;

    pub use kernel::prelude::*;

    pub fn word(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
    }

    pub fn spaces(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| " \t".contains(c))(i)
    }

    pub fn noun(i: &str) -> IResult<&str, Item> {
        map(word, |s: &str| Item::Named(s.to_owned()))(i)
    }

    pub fn text_to_end_of_line(i: &str) -> IResult<&str, &str> {
        take_while1(|_c: char| true)(i)
    }

    pub fn string_literal(i: &str) -> IResult<&str, &str> {
        delimited(tag("\""), string_inside, tag("\""))(i)
    }

    fn string_inside(i: &str) -> IResult<&str, &str> {
        take_while(|c: char| c.is_alphabetic() || c.is_whitespace())(i)
    }

    pub fn camel_case_word(i: &str) -> IResult<&str, &str> {
        word(i) // TODO
    }

    pub fn unsigned_number(i: &str) -> IResult<&str, u64> {
        map_res(recognize(digit1), str::parse)(i)
    }

    pub fn gid_reference(i: &str) -> IResult<&str, Item> {
        map(preceded(tag("#"), unsigned_number), |n| {
            Item::Gid(EntityGid::new(n))
        })(i)
    }

    pub fn surrounding_area(i: &str) -> IResult<&str, Item> {
        map(alt((tag("area"), tag("here"))), |_s: &str| Item::Area)(i)
    }

    pub fn noun_or_specific(i: &str) -> IResult<&str, Item> {
        alt((surrounding_area, myself, noun, gid_reference))(i)
    }

    pub fn myself(i: &str) -> IResult<&str, Item> {
        map(alt((tag("self"), tag("myself"))), |_s: &str| Item::Myself)(i)
    }

    pub fn named_place(i: &str) -> IResult<&str, Item> {
        alt((
            gid_reference,
            map(word, |s: &str| Item::Route(s.to_owned())),
        ))(i)
    }

    pub fn try_parsing<T: ParsesActions>(parser: T, i: &str) -> EvaluationResult {
        parser.try_parse_action(i)
    }
}

pub mod model {
    pub use anyhow::Result;
    pub use chrono::{DateTime, Utc};
    pub use kernel::common::*;
    pub use kernel::prelude::*;
    pub use macros::*;
    pub use serde::{Deserialize, Serialize};
    pub use serde_json::json;
    pub use std::rc::Rc;
    pub use std::{collections::HashMap, ops::Deref};
    pub use tracing::*;
}

pub mod actions {
    pub use crate::tools;
    pub use anyhow::Result;
    pub use kernel::common::*;
    pub use kernel::prelude::*;
    pub use macros::*;
    pub use serde::{Deserialize, Serialize};
    pub use tracing::*;

    pub fn reply_ok<T: ToTaggedJson + 'static>(
        living: EntityPtr,
        audience: Audience,
        raise: T,
    ) -> Result<Effect> {
        get_my_session()?.raise(
            Some(living),
            audience,
            Raising::TaggedJson(raise.to_tagged_json()?),
        )?;

        Ok(Effect::Ok)
    }

    pub fn reply_done<T: ToTaggedJson + 'static>(
        living: EntityPtr,
        audience: Audience,
        raise: T,
    ) -> Result<SimpleReply> {
        get_my_session()?.raise(
            Some(living),
            audience,
            Raising::TaggedJson(raise.to_tagged_json()?),
        )?;

        Ok(SimpleReply::Done)
    }
}

pub mod tests {
    pub use anyhow::Result;
    pub use chrono::TimeZone;
    pub use chrono::Utc;
    pub use serde::de::DeserializeOwned;
    pub use std::collections::HashSet;
    pub use std::rc::Rc;

    pub use crate::tools;
    pub use crate::{BuildSurroundings, QuickThing};
    pub use kernel::common::*;
    pub use kernel::here;
    pub use kernel::prelude::*;
    pub use tracing::*;

    pub use super::plugin::try_parsing;

    pub trait ToDebugJson {
        fn to_debug_json(&self) -> Result<JsonValue, serde_json::Error>;
    }

    impl ToDebugJson for Effect {
        fn to_debug_json(&self) -> Result<JsonValue, serde_json::Error> {
            serde_json::to_value(self)
        }
    }

    pub fn parse_and_perform<T: ParsesActions>(
        parser: T,
        line: &str,
    ) -> Result<(Surroundings, Effect)> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.plain().encyclopedia()?.build()?;
        let action = try_parsing(parser, line)?;
        let action = action.unwrap();
        let effect = action.perform(session.clone(), &surroundings)?;
        build.close()?;
        Ok((surroundings, effect))
    }

    pub fn perform_directly<T: Action>(action: T) -> Result<(Surroundings, Effect)> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.plain().encyclopedia()?.build()?;
        let action: Rc<dyn Action> = Rc::new(action);
        let effect = action.perform(session.clone(), &surroundings)?;
        build.close()?;
        Ok((surroundings, effect))
    }
}
