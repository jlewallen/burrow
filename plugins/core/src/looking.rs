use crate::library::plugin::*;

#[derive(Default)]
pub struct LookingPluginFactory {}

impl PluginFactory for LookingPluginFactory {
    fn create_plugin(&self) -> Result<Box<dyn Plugin>> {
        Ok(Box::new(LookingPlugin {}))
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

#[derive(Default)]
pub struct LookingPlugin {}

impl Plugin for LookingPlugin {
    fn plugin_key() -> &'static str
    where
        Self: Sized,
    {
        "looking"
    }

    fn key(&self) -> &'static str {
        Self::plugin_key()
    }

    fn initialize(&mut self) -> Result<()> {
        Ok(())
    }

    fn register_hooks(&self, _hooks: &ManagedHooks) -> Result<()> {
        Ok(())
    }

    fn have_surroundings(&self, _surroundings: &Surroundings) -> Result<()> {
        Ok(())
    }

    fn deliver(&self, _incoming: &Incoming) -> Result<()> {
        Ok(())
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

impl ParsesActions for LookingPlugin {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        try_parsing(parser::LookActionParser {}, i)
    }
}

pub mod model {
    use thiserror::Error;

    use crate::library::model::*;
    use crate::{
        carrying::model::{Carryable, Containing},
        moving::model::{Movement, Occupyable},
    };

    pub fn qualify_name(quantity: f32, name: &str) -> String {
        use indefinite::*;
        use inflection::*;
        if quantity > 1.0 {
            let pluralized = plural::<_, String>(name);
            format!("{} {}", quantity, &pluralized)
        } else {
            indefinite(name)
        }
    }

    pub trait ObserveHook<T> {
        fn observe(
            &self,
            surroundings: &Surroundings,
            user: &Entry,
            target: &Entry,
        ) -> Result<Option<T>>;
    }

    pub trait Observe<T> {
        fn observe(&self, user: &Entry) -> Result<Option<T>>;
    }

    impl Observe<ObservedEntity> for &Entry {
        fn observe(&self, _user: &Entry) -> Result<Option<ObservedEntity>> {
            let name = self.name()?;
            let carryable = self.scope::<Carryable>()?;
            let qualified = name.as_ref().map(|n| qualify_name(carryable.quantity(), n));

            Ok(Some(ObservedEntity {
                key: self.key().to_string(),
                name,
                qualified,
                desc: self.desc()?,
            }))
        }
    }

    pub fn new_entity_observation(
        user: &Entry,
        entity: &Entry,
    ) -> Result<Option<EntityObservation>> {
        Ok(entity
            .observe(user)?
            .map(|entity| EntityObservation { entity }))
    }

    pub fn new_inside_observation(
        user: &Entry,
        vessel: &Entry,
    ) -> Result<Option<InsideObservation>> {
        let mut items = Vec::new();
        if let Ok(containing) = vessel.scope::<Containing>() {
            for lazy_entity in &containing.holding {
                let entity = &lazy_entity.into_entry()?;
                items.push(entity.observe(user)?);
            }
        }

        Ok(vessel.observe(user)?.map(|vessel| InsideObservation {
            vessel,
            items: items.into_iter().flatten().collect(),
        }))
    }

    pub fn new_area_observation(user: &Entry, area: &Entry) -> Result<AreaObservation> {
        let mut living: Vec<ObservedEntity> = vec![];
        if let Ok(occupyable) = area.scope::<Occupyable>() {
            for entity in &occupyable.occupied {
                if let Some(observed) = (&entity.into_entry()?).observe(user)? {
                    living.push(observed);
                }
            }
        }

        let mut items = vec![];
        if let Ok(containing) = area.scope::<Containing>() {
            for entity in &containing.holding {
                items.push((&entity.into_entry()?).observe(user)?);
            }
        }

        let mut carrying = vec![];
        if let Ok(containing) = user.scope::<Containing>() {
            for entity in &containing.holding {
                carrying.push((&entity.into_entry()?).observe(user)?);
            }
        }

        let mut routes = vec![];
        if let Ok(movement) = user.scope::<Movement>() {
            for route in &movement.routes {
                routes.push((&route.area.into_entry()?).observe(user)?);
            }
        }

        Ok(AreaObservation {
            area: area
                .observe(user)?
                .ok_or_else(|| LookError::InvisibleSurroundingArea)?,
            person: user
                .observe(user)?
                .ok_or_else(|| LookError::InvisibleSelf)?,
            living,
            items: items.into_iter().flatten().collect(),
            carrying: carrying.into_iter().flatten().collect(),
            routes: routes.into_iter().flatten().collect(),
        })
    }

    #[derive(Error, Debug)]
    pub enum LookError {
        #[error("Invisible surrounding area")]
        InvisibleSurroundingArea,
        #[error("Invisible self")]
        InvisibleSelf,
    }
}

pub mod actions {
    use anyhow::Context;

    use super::model::*;
    use crate::library::actions::*;

    #[derive(Debug)]
    pub struct LookAction {}

    impl Action for LookAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, _session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            let (_, user, area) = surroundings.unpack();

            Ok(new_area_observation(&user, &area)
                .with_context(|| "Observing area")?
                .into())
        }
    }

    #[derive(Debug)]
    pub struct LookInsideAction {
        pub item: Item,
    }

    impl Action for LookInsideAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            let (_, user, _area) = surroundings.unpack();

            match session.find_item(surroundings, &self.item)? {
                Some(target) => {
                    if tools::is_container(&target)? {
                        match new_inside_observation(&user, &target)? {
                            Some(observation) => Ok(observation.into()),
                            None => Ok(SimpleReply::NotFound.into()),
                        }
                    } else {
                        Ok(SimpleReply::Impossible.into())
                    }
                }
                None => Ok(SimpleReply::NotFound.into()),
            }
        }
    }

    #[derive(Debug)]
    pub struct LookAtAction {
        pub item: Item,
    }

    impl Action for LookAtAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            let (_, user, _area) = surroundings.unpack();

            match session.find_item(surroundings, &self.item)? {
                Some(target) => match new_entity_observation(&user, &target)? {
                    Some(observation) => Ok(observation.into()),
                    None => Ok(SimpleReply::NotFound.into()),
                },
                None => Ok(SimpleReply::NotFound.into()),
            }
        }
    }
}

pub mod parser {
    use crate::library::parser::*;

    use super::actions::{LookAction, LookAtAction, LookInsideAction};

    pub struct LookActionParser {}

    impl ParsesActions for LookActionParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let inside = map(
                separated_pair(
                    separated_pair(tag("look"), spaces, tag("inside")),
                    spaces,
                    noun,
                ),
                |(_, nearby)| Box::new(LookInsideAction { item: nearby }) as Box<dyn Action>,
            );

            let at = map(
                separated_pair(separated_pair(tag("look"), spaces, tag("at")), spaces, noun),
                |(_, nearby)| Box::new(LookAtAction { item: nearby }) as Box<dyn Action>,
            );

            let area = map(tag("look"), |_| Box::new(LookAction {}) as Box<dyn Action>);

            let (_, action) = alt((inside, at, area))(i)?;

            Ok(action)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::model::*;
    use super::parser::LookActionParser;
    use super::*;
    use crate::{library::plugin::try_parsing, BuildSurroundings, QuickThing};

    #[test]
    fn it_looks_in_empty_area() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.plain().build()?;

        let action = try_parsing(LookActionParser {}, "look")?;
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_looks_in_area_with_items_on_ground() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build
            .ground(vec![QuickThing::Object("Cool Rake")])
            .ground(vec![QuickThing::Object("Boring Shovel")])
            .build()?;

        let action = try_parsing(LookActionParser {}, "look")?;
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_looks_in_area_with_items_on_ground_and_a_route() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let destination = build.make(QuickThing::Place("Place"))?;
        let (session, surroundings) = build
            .ground(vec![QuickThing::Object("Cool Rake")])
            .ground(vec![QuickThing::Object("Boring Shovel")])
            .route("East Exit", QuickThing::Actual(destination))
            .build()?;

        let action = try_parsing(LookActionParser {}, "look")?;
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_looks_in_area_with_items_on_ground_and_holding_items() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let destination = build.make(QuickThing::Place("Place"))?;
        let (session, surroundings) = build
            .ground(vec![QuickThing::Object("Boring Shovel")])
            .hands(vec![QuickThing::Object("Cool Rake")])
            .route("East Exit", QuickThing::Actual(destination))
            .build()?;

        let action = try_parsing(LookActionParser {}, "look")?;
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_fails_to_look_inside_non_containers() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.hands(vec![QuickThing::Object("Not A Box")]).build()?;

        let action = try_parsing(LookActionParser {}, "look inside box")?;
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_looks_inside_containers() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let vessel = build
            .entity()?
            .named("Vessel")?
            .holding(&vec![build.make(QuickThing::Object("Key"))?])?
            .into_entry()?;
        let (session, surroundings) = build.hands(vec![QuickThing::Actual(vessel)]).build()?;

        let action = try_parsing(LookActionParser {}, "look inside vessel")?;
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_fails_to_look_at_not_found_entities() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let vessel = build.entity()?.named("Hammer")?.into_entry()?;
        let (session, surroundings) = build.hands(vec![QuickThing::Actual(vessel)]).build()?;

        let action = try_parsing(LookActionParser {}, "look at shovel")?;
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_looks_at_entities() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let vessel = build.entity()?.named("Hammer")?.into_entry()?;
        let (session, surroundings) = build.hands(vec![QuickThing::Actual(vessel)]).build()?;

        let action = try_parsing(LookActionParser {}, "look at hammer")?;
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn qualify_name_basics() {
        // Not going to test all of indefinite's behavior here, just build edge
        // cases in our integrating logic.
        assert_eq!(qualify_name(1.0, "box"), "a box");
        assert_eq!(qualify_name(2.0, "box"), "2 boxes");
        assert_eq!(qualify_name(1.0, "person"), "a person");
        assert_eq!(qualify_name(2.0, "person"), "2 people");
        assert_eq!(qualify_name(1.0, "orange"), "an orange");
        assert_eq!(qualify_name(2.0, "orange"), "2 oranges");
        assert_eq!(qualify_name(1.0, "East Exit"), "an East Exit");
    }
}
