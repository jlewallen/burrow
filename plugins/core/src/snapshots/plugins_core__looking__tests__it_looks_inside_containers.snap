---
source: plugins/core/src/looking.rs
expression: reply.to_json()?
---
{
  "basicReply": {
    "insideObservation": {
      "items": [
        {
          "desc": null,
          "key": "E-1",
          "name": "Key",
          "qualified": "a Key"
        }
      ],
      "vessel": {
        "desc": null,
        "key": "E-0",
        "name": "Vessel",
        "qualified": "a Vessel"
      }
    }
  }
}
